src/common/Finisher.h

```diff
@@ -67,58 +70,59 @@ class Finisher {
   /// Add a context to complete, optionally specifying a parameter for
   /// the complete function.
   void queue(Context *c, int r = 0) {
-    std::unique_lock ul(finisher_lock);
+    std::scoped_lock l(finisher_lock);
     if (finisher_queue.empty()) {
       finisher_cond.notify_all();
     }
-    finisher_queue.push_back(make_pair(c, r));
+
+    finisher_queue.emplace_back([c, r]() noexcept {c->complete(r); });
```

Comment:
Maybe I'm being paranoid, but I worry this is adding an additional layer of overhead above and on top of the Context

Reply:
It's adding an additional layer, but the amount of overhead should be minimal.

The lambda itself uses two machine words which is basically what we'd use anyway for storing a context and a return code.

std::function /may/ depending on the implementation inline that. (I know libcxx does.) (If we pull in cxx_funcion::unique_function and use that instead I can guarantee that it will be inlined.)

The call through std::function /may/ be a virtual function call. (Or may not. libcxx IS, or at least was last time I looked). unique_function definitely isn't.

If you prefer I can have the queue hold a std::variant which can also guarantee inlining and no virtual function overhead.

(There's a run-time check of the type-tag but that shouldn't be much.)

Reply:
Maybe we can do a quick microbenchmark as part of the unit test?  Queue and fire 100,000 trivial Contexts and lambda and just make sure it isn't introducing a regression we'll have a hard time remembering/finding later?

Reply:
All right! I shall throw in a benchmark and see what I get.

Reply:
All right! Sorry for the delay, I got pulled into doing Stuff:

BUT! With a queue and completion of 1,000,000 contexts with Release build type, I get:

Original code: 0.0500009s
New code: 0.055001s

New code is, perhaps, very slightly slower? Having run them both a few times the variability between runs is wider than the variability between the different implementations so it could just be noise. Either way, the new version that turns Context into a lambda doesn't seem to hurt performance.

Ether way, I could slap an EXPECT_LT on it, but I'm a bit loath to do that because I don't want to introduce something that will spuriously fail on a loaded machine or if compiled with -O0 (as I usually do when debugging.)

(You can see the code for the test in commit 7287ab67d9b8d74daed24b67acf6946c8f5fb584 in case you want to check I didn't miss something obvious.)